File: lab4-records/java/Playlist.java
public class Playlist {

    public static class Person {
        public final String name;

        public Person(String name) {
            this.name = name;
        }

        @Override
        public String toString() {
            return name;
        }

    }

    public static class Item {
        public final String name;
        public final Person performer;
        public final Float length_secs;

        public Item(String name, Playlist.Person performer, Float length_secs) {
            this.name = name;
            this.performer = performer;
            this.length_secs = length_secs;
        }

        @Override
        public String toString() {
            return String.format("%s by %s (%.1fs)", name, performer.toString(), length_secs);
        }
    }

    private static Item piece1 = new Item(
            "Moonlight Sonata",
            new Person("Claudio Arrau"),
            17 * 60 + 26f);

    // private static Item pause1 = new Pause(5f);

    public static void main(String[] args) {
        System.out.println(piece1);
        // System.out.println(pause1);
    }
}

File: lab4-records/haskell/playlist.hs
module Main where

import Text.Printf

-- start of code to focus on in Practical 4

data Person
    = Person
        {
          person_name :: String
        }
    deriving (Eq)

instance Show Person where 
    -- using Java terminology: Person implements interface Show
    show (Person name) = name -- how to format a Person record

data Item
    = Piece
        {
            item_name :: String,
            item_performer :: Person,
            item_length_secs :: Float
        }
    deriving (Eq)

instance (Show Item) where
    show (Piece name performer len) =
        printf "%s by %s (%.1fs)" name (show performer) len

piece1 =
    Piece
    { 
        item_name = "Moonlight Sonata",
        item_performer = Person "Claudio Arrau",
        item_length_secs = 17*60+26
    }

piece2 =
    Piece
    { 
        item_name = "Moonlight Sonata",
        item_performer = Person "Daniel Barenboim",
        item_length_secs = 16*60+49
    }
  
{-
pause1 =
    Pause
    { 
        item_length_secs = 5
    }
-}

main =
    do
    -- putStrLn "piece1 and piece2 sorted by length:"
    -- putStrLn $ show shorterPiece
    -- putStrLn $ show longerPiece
    putStr "piece1 = "
    putStrLn $ show piece1
--    putStr "pause1 = "
--    putStrLn $ show pause1

-- ... = sortTwoItems (piece1, piece2) -- TASK

sortTwoItems (item1, item2) = 
    if item_length_secs item1 <= item_length_secs item2
        then (item1, item2)
        else (item2, item1)

